gk/main.c

```diff
@@ -16,22 +16,404 @@
+{
+	bool renew_cap;
+	uint64_t now = rte_rdtsc();
+	struct rte_mbuf *pkt = packet->pkt;
+
+	/* 
+	 * When a flow entry is at granted state, the GK block processes 
+	 * that entry as follows:
+	 */
+	if (now >= fe->u.granted.cap_expire_at) {
+		initialize_flow_entry(fe, &packet->flow);
+		return gk_process_request(fe, packet);
+	}
+
+	if (now >= fe->u.granted.budget_renew_at) {
+		fe->u.granted.budget_renew_at = now + time_from_seconds(1);
+		fe->u.granted.budget_byte = fe->u.granted.tx_rate_kb_cycle * 1024;
+	}
+
+	if (pkt->data_len > fe->u.granted.budget_byte)
+		return drop_packet(pkt);
+
+	fe->u.granted.budget_byte -= pkt->data_len;
+	renew_cap = now >= fe->u.granted.send_next_renewal_at;
+	if (renew_cap)
+		fe->u.granted.send_next_renewal_at = now +
+			fe->u.granted.renewal_step_cycle;
+
+	/*
+	 * TODO Encapsulate packet as a granted packet,
+	 * mark it as a capability renewal request if @renew_cap is true,
+	 * enter destination according to @fe->u.granted.grantor_id,
+	 * and put the encapsulated packet in the granted queue.
+	 */
+
+	return 0;
+}
+
+static int
+gk_process_declined(struct flow_entry *fe, struct ipacket *packet)
+{
+	uint64_t now = rte_rdtsc();
+
+	/*
+	 * When a flow entry is at declined state, the GK block processes
+	 * that entry as follows:
+	 */
+	if (unlikely(now >= fe->u.declined.expire_at)) {
+		initialize_flow_entry(fe, &packet->flow);
+		return gk_process_request(fe, packet);
+	}
+
+	return drop_packet(packet->pkt);
+}
+
+static int
+setup_gk_instance(unsigned int lcore_id, struct gk_config *gk_conf)
+{
+	int  ret;
+	char ht_name[64];
+	unsigned int block_idx = lcore_id - gk_conf->lcore_start_id;
+
+	struct gk_instance *instance = &gk_conf->instances[block_idx];
+	struct rte_hash_parameters ip_flow_hash_params = {
+		.name = NULL,
+		.entries = gk_conf->flow_ht_size,
+		.key_len = sizeof(struct ip_flow),
+		.hash_func = DEFAULT_HASH_FUNC,
+		.hash_func_init_val = 0,
+	};
+
+	ret = snprintf(ht_name, sizeof(ht_name), "ip_flow_hash_%u", block_idx);
+	RTE_ASSERT(ret < sizeof(ht_name));
+
+	/* Setup the flow hash table for GK block @block_idx. */
+	ip_flow_hash_params.name = ht_name;
+	ip_flow_hash_params.socket_id = rte_lcore_to_socket_id(lcore_id);
+	instance->ip_flow_hash_table = rte_hash_create(&ip_flow_hash_params);
+	if (!instance->ip_flow_hash_table) {
+		RTE_LOG(ERR, HASH,
+			"The GK block cannot create hash table at lcore %u!\n",
+			lcore_id);
+
+		ret = -1;
+		goto out;
+	}
+
+	/* Setup the flow entry table for GK block @block_idx. */
+	instance->ip_flow_entry_table = (struct flow_entry *)rte_calloc(NULL,
+		gk_conf->flow_ht_size, sizeof(struct flow_entry), 0);
+	if (!instance->ip_flow_entry_table) {
+		RTE_LOG(ERR, MALLOC,
+			"The GK block can't create flow entry table at lcore %u!\n",
+			lcore_id);
+
+		ret = -1;
+		goto flow_hash;
+	}
+
+	goto out;
+
+flow_hash:
+	rte_hash_free(instance->ip_flow_hash_table);
+	instance->ip_flow_hash_table = NULL;
+out:
+	return (ret == -1 ? -1 : 0);
```

Comment:
Drop this test and just set `ret = 0` before the last `goto out`. This test is just making the code harder to read.